src/Toml/Parser.hs

```diff
@@ -1,13 +1,184 @@
 module Toml.Parser
-       ( parse
+       ( ParseException (..)
+       , parse
        ) where
 
+-- I hate default Prelude... Do I really need to import all this stuff manually?..
+import Control.Applicative (Alternative (..), liftA2)
+import Control.Applicative.Combinators (between, manyTill, sepBy)
+import Control.Monad (void)
+import Data.Function (on)
+import Data.HashMap.Lazy (HashMap)
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Ord (comparing)
+import Data.Semigroup ((<>))
 import Data.Text (Text)
+import Data.Void (Void)
+import Text.Megaparsec (Parsec, parseErrorPretty', try)
+import Text.Megaparsec.Char (alphaNumChar, anyChar, char, space, space1)
 
-import Toml.Type (TOML)
+import Toml.Type (AnyValue, Key (..), TOML (..), UValue (..), typeCheck)
 
--- | Dummy parse exception for now.
-data ParseException = ParseException
+import qualified Control.Applicative.Combinators.NonEmpty as NC
+import qualified Data.HashMap.Lazy as HashMap
+import qualified Data.List as List (isPrefixOf, sortBy)
+import qualified Data.Text as Text
+import qualified Text.Megaparsec as Mega (parse)
+import qualified Text.Megaparsec.Char.Lexer as L
 
+----------------------------------------------------------------------------
+-- Library for parsing (boilerplate copy-pasted from megaparsec tutorial)
+----------------------------------------------------------------------------
+
+type Parser = Parsec Void Text
+
+-- space consumer
+sc :: Parser ()
+sc = L.space space1 lineComment blockComment
+  where
+    lineComment  = L.skipLineComment "#"
+    blockComment = empty
+
+-- wrapper for consuming spaces after every lexeme (not before it!)
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme sc
+
+-- parser for "fixed" string
+text :: Text -> Parser Text
+text = L.symbol sc
+
+text_ :: Text -> Parser ()
+text_ = void . text
+
+integer :: Parser Integer
+integer = L.signed sc (lexeme L.decimal)
+
+float :: Parser Double
+float = lexeme L.float
+
+----------------------------------------------------------------------------
+-- TOML parser
+----------------------------------------------------------------------------
+
+-- Keys
+
+bareKeyP :: Parser Text
+bareKeyP = lexeme $ Text.pack <$> bareStrP
+  where
+    bareStrP :: Parser String
+    bareStrP = some $ alphaNumChar <|> char '_' <|> char '-'
+
+stringP :: Parser Text
+stringP = lexeme $ Text.pack <$> (char '"' *> anyChar `manyTill` char '"')
```

Comment:
Well, I'm not sure how to write such parser properly... Because `'` is for literal strings. With `'` I can write:

```toml
quoted   = 'Tom "Dubs" Preston-Werner'
```

But with `"` this needed to be write like this:

```toml
quoted   = "Tom \"Dubs\" Preston-Werner"
```

Actually, I'm not sure which behavior is currently implemented and what be the result of this parser... In `""` strings I also need to support this:

```
\b         - backspace       (U+0008)
\t         - tab             (U+0009)
\n         - linefeed        (U+000A)
\f         - form feed       (U+000C)
\r         - carriage return (U+000D)
\"         - quote           (U+0022)
\\         - backslash       (U+005C)
\uXXXX     - unicode         (U+XXXX)
\UXXXXXXXX - unicode         (U+XXXXXXXX)
```

Not sure how to do this easilty and what comes out of the box...

Reply:
Okay, `"Tom \"Dubs\" Preston-Werner"` is not even parseable because current parser interprets `\` as a separate character which makes sense. So I just need to replace `"` with `'` to have parser of literal strings.